grilops/regions.py

```
@@ -143,54 +150,36 @@ def subtree_size_term(sp, sd):
           0
       )
 
-    for p in self.__locations:
+    for p in self.__lattice.points:
       parent = self.__parent_grid[p]
       subtree_size_terms = [
-          If(parent != X, 1, 0)
+          If(parent != self.parent_type_to_index("X"), 1, 0)
       ]
 
-      sp = p.translate(Vector(-1, 0))
-      if sp in self.__parent_grid:
-        constrain_side(p, sp, S)
-        subtree_size_terms.append(subtree_size_term(sp, S))
-      else:
-        self.__solver.add(parent != N)
-
-      sp = p.translate(Vector(1, 0))
-      if sp in self.__parent_grid:
-        constrain_side(p, sp, N)
-        subtree_size_terms.append(subtree_size_term(sp, N))
-      else:
-        self.__solver.add(parent != S)
-
-      sp = p.translate(Vector(0, -1))
-      if sp in self.__parent_grid:
-        constrain_side(p, sp, E)
-        subtree_size_terms.append(subtree_size_term(sp, E))
-      else:
-        self.__solver.add(parent != W)
-
-      sp = p.translate(Vector(0, 1))
-      if sp in self.__parent_grid:
-        constrain_side(p, sp, W)
-        subtree_size_terms.append(subtree_size_term(sp, W))
-      else:
-        self.__solver.add(parent != E)
+      for d in self.__adjacency_directions:
+        sp = p.translate(d)
+        if sp in self.__parent_grid:
+          opposite_index = self.__adjacency_to_index[d.negate()]
+          constrain_side(p, sp, opposite_index)
+          subtree_size_terms.append(subtree_size_term(sp, opposite_index))
+        else:
+          d_index = self.__adjacency_to_index[d]
+          self.__solver.add(parent != d_index)
 
       self.__solver.add(
           self.__subtree_size_grid[p] == Sum(*subtree_size_terms)
       )
 
-  def location_to_region_id(self, location: Point) -> int:
+  def location_to_region_id(self, location: Point) -> Optional[int]:
```

maybe this should raise instead of returning None if the point's not in the lattice?

---

Done!